systemd/manager.go

```diff
@@ -196,17 +196,25 @@ func (m *systemdUnitManager) GetUnitStates(filter pkg.Set) (map[string]*unit.Uni
 	// operations could mutate the hashes before we've retrieved the state
 	// for every unit in the filter, since they won't necessarily all be
 	// present in the initial ListUnits() call.
+	fallback := false
+
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-	dbusStatuses, err := m.systemd.ListUnits()
+	dbusStatuses, err := m.systemd.ListUnitsByNames(filter.Values())
 
 	if err != nil {
-		return nil, err
+		fallback = true
+		log.Warningf("ListUnitsByNames is not implemented in your systemd version, fallback to ListUnits: %v", err)
+		dbusStatuses, err = m.systemd.ListUnits()
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	states := make(map[string]*unit.UnitState)
 	for _, dus := range dbusStatuses {
-		if !filter.Contains(dus.Name) {
+		if fallback && !filter.Contains(dus.Name) {
```

Comment:
Reasonable :) thanks. I refactored this patch several times and didn't notice that.